### PR TITLE
Use GitHub actions for unit testing

### PR DESCRIPTION
**Pull request by peanutfun**
Changes proposed in this PR:
- Define allowed Python versions in `setup.py` and environment file
- Allow Python 3.10 and 3.11 for Climada
- Add a GitHub workflow for CI
   - Build new conda environments for every branch. Environments are cached over one calendar day, meaning that they do not have to be rebuilt during that period if the environment files don't change.
   - Test CLIMADA with Python 3.9, 3.10, 3.11
   - Run unit tests and report test results
   - Upload coverage reports as artifacts
- Fix some tests
  - Two nightlight tests are actually integration tests, as they require that some files have been downloaded
  - Adjust accuracy on some tests which fail otherwise
  - Use `default_factory` to initialize mutable default values for dataclass

### PR Author Checklist

- [x] Read the [Contribution Guide][contrib]
- [x] Correct target branch selected (if unsure, select `develop`)
- [x] Descriptive pull request title added
- [x] Source branch up-to-date with target branch
- [ ] [Documentation](https://climada-python.readthedocs.io/en/latest/guide/Guide_PythonDos-n-Donts.html#2.--Commenting-&-Documenting) updated
- [x] [Tests][testing] updated
- [x] [Tests][testing] passing
- [x] No new [linter issues][linter]
- [ ] [Changelog](https://github.com/CLIMADA-project/climada_python/blob/main/CHANGELOG.md) updated

### PR Reviewer Checklist

- [ ] Read the [Contribution Guide][contrib]
- [ ] [CLIMADA Reviewer Checklist](https://climada-python.readthedocs.io/en/latest/guide/Guide_Reviewer_Checklist.html) passed
- [ ] [Tests][testing] passing
- [ ] No new [linter issues][linter]

[contrib]: https://github.com/CLIMADA-project/climada_python/blob/main/CONTRIBUTING.md
[testing]: https://climada-python.readthedocs.io/en/latest/guide/Guide_Continuous_Integration_and_Testing.html
[linter]: https://climada-python.readthedocs.io/en/stable/guide/Guide_Continuous_Integration_and_Testing.html#3.C.--Static-Code-Analysis

### CODE REVIEW

**Comment by peanutfun** [comment]
I tried to achieve compatibility also with Python 3.11, but the unsequa module causes segmentation faults there, see https://github.com/CLIMADA-project/climada_python/actions/runs/5890648735/job/15976187422. I think this is due to the process pool. Maybe #763 will resolve these issues, but we will have to look out.

**Comment by chahank** [comment]
The build in 3.10 also fails, but because of another module. Can this be easily fixed?

**Comment by peanutfun** [comment]
All checks have passed for the latest commit. What are you referring to? 🤔

**Comment by peanutfun** [comment]
This was the test result of an earlier commit. The particular error was resolved by https://github.com/CLIMADA-project/climada_python/pull/745/commits/109d99443b5f72498e3291b42535eb3e6571ce13

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    name: Build Env, Install, Unit Tests
+    runs-on: ubuntu-latest
+    permissions:
+      # For publishing results
+      checks: write
+
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: ["3.9", "3.10"]
```
Comment (chahank) [inline .github/workflows/ci.yml:21]: Is this something we then have to update manually for each release?

**Comment by chahank** [comment]
Looks good to me if this does what it should do which I cannot tell from looking at the code only. Maybe @emanuel-schmid can confirm?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (peanutfun) [reply .github/workflows/ci.yml:21]: For each *Python* release, yes.

**Comment by peanutfun** [comment]
@chahank @emanuel-schmid How and where should I document this? (topics: What does the pipeline do, how to inspect results and coverage reports, etc)

**Comment by peanutfun** [comment]
> I tried to achieve compatibility also with Python 3.11, but the unsequa module causes segmentation faults there, see https://github.com/CLIMADA-project/climada_python/actions/runs/5890648735/job/15976187422. I think this is due to the process pool.

This apparently was resolved by #763. Excellent work, @chahank!

**Comment by peanutfun** [comment]
> @chahank @emanuel-schmid How and where should I document this? (topics: What does the pipeline do, how to inspect results and coverage reports, etc)

I would highly appreciate your comments on this. Apart from that, the branch is ready to be merged.

**Comment by emanuel-schmid** [comment]
Well - there is https://github.com/CLIMADA-project/climada_python/blob/main/doc/guide/Guide_Continuous_Integration_and_Testing.ipynb where it wouldn't be completely out of place. 
But atm I'd suggest to keep it basic - let's see whether it's sustainable before making a full-fledged documentation about it.

**Comment by peanutfun** [comment]
From personal discussion: I'll add a documentation of the pipeline itself but not make it part of the guides for testing yet.

**Pull request merged**